File: 86-partition-list/partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
    partOne := &ListNode{}
    partTwo := &ListNode{}

    pOne := partOne
    pTwo := partTwo

    for head != nil {
        if head.Val < x {
            pOne.Next = head
            pOne = pOne.Next
        } else {
            pTwo.Next = head
            pTwo = pTwo.Next
        }

        head = head.Next
    }

    pTwo.Next = nil
    pOne.Next = partTwo.Next

    return partOne.Next
}

File: 86-partition-list/partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* part1 = new ListNode();
        ListNode* part2 = new ListNode();
        ListNode* p1 = part1;
        ListNode* p2 = part2;

        while(head) {
            if (head->val < x) {
                p1->next = head;
                p1 = p1->next;
            } else {
                p2->next = head;
                p2 = p2->next;
            }

            head = head->next;
        }

        p2->next = nullptr;
        p1->next = part2->next;
        return part1->next;
    }
};